core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java

```diff
@@ -0,0 +1,509 @@
+    this.regionId = connFields[1];
+    this.instanceId = connFields[2];
+
+    this.apiClient = apiClient;
+    this.executor = executor;
+    this.keyPair = keyPair;
+
+    // Kick off initial async jobs
+    synchronized (instanceDataGuard) {
+      this.currentInstanceData = performRefresh();
+      this.nextInstanceData = Futures.immediateFuture(currentInstanceData);
+    }
+  }
+
+  /**
+   * Returns the current data related to the instance from {@link #performRefresh()}. May block if
+   * no valid data is currently available.
+   */
+  private InstanceData getInstanceData() {
+    ListenableFuture<InstanceData> instanceData;
+    synchronized (instanceDataGuard) {
+      instanceData = currentInstanceData;
+    }
+    try {
+      // TODO(kvg): Let exceptions up to here before adding context
+      return Uninterruptibles.getUninterruptibly(instanceData);
+    } catch (ExecutionException ex) {
+      Throwable cause = ex.getCause();
+      Throwables.throwIfUnchecked(cause);
+      throw new RuntimeException(cause);
+    }
+  }
+
+  /**
+   * Returns an unconnected {@link SSLSocket} using the SSLContext associated with the instance. May
+   * block until required instance data is available.
+   */
+  SSLSocket createSslSocket() throws IOException {
+    return (SSLSocket) getInstanceData().getSslContext().getSocketFactory().createSocket();
+  }
+
+  /**
+   * Returns the first IP address for the instance, in order of the preference supplied by
+   * preferredTypes.
+   *
+   * @param preferredTypes Preferred instance IP types to use. Valid IP types include "Public" and
+   *     "Private".
+   * @return returns a string representing the IP address for the instance
+   * @throws IllegalArgumentException If the instance has no IP addresses matching the provided
+   *     preferences.
+   */
+  String getPreferredIp(List<String> preferredTypes) {
+    Map<String, String> ipAddrs = getInstanceData().getIpAddrs();
+    for (String ipType : preferredTypes) {
+      String preferredIp = ipAddrs.get(ipType);
+      if (preferredIp != null) {
+        return preferredIp;
+      }
+    }
+    throw new IllegalArgumentException(
+        String.format(
+            "[%s] Cloud SQL instance  does not have any IP addresses matching preferences (%s)",
+            connectionName, String.join(", ", preferredTypes)));
+  }
+
+  /**
+   * Attempts to force a new refresh of the instance data. May fail if called too frequently or if a
+   * new refresh is already in progress. If successful, other methods will block until refresh has
+   * been completed.
+   *
+   * @return {@code true} if successfully scheduled, or {@code false} otherwise.
+   */
+  boolean forceRefresh() {
+    if (!forcedRenewRateLimiter.tryAcquire()) {
+      return false;
+    }
+    synchronized (instanceDataGuard) {
+      // If a scheduled refresh doesn't exist or hasn't started, perform one immediately
+      if (nextInstanceData == null || nextInstanceData.cancel(false)) {
```

Comment:
I don't think `nextInstanceData` is ever `null` now.

```suggestion
      if (nextInstanceData.cancel(false)) {
```